Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx

```diff
@@ -0,0 +1,351 @@
+    for (int i = 0; i < chambers.length(); ++i) {
+      if (chambers[i] == ',') {
+        continue;
+      }
+      int chamber = chambers[i] - '0';
+      LOG(info) << Form("%s%d", "Fixing chamber: ", chamber);
+      mAlign.FixChamber(chamber);
+    }
+
+    // Init for output saving
+    auto OutputRecFileName = ic.options().get<string>("output-record-data");
+    auto OutputConsFileName = ic.options().get<string>("output-record-constraint");
+    mAlign.init(OutputRecFileName, OutputConsFileName);
+
+    ic.services().get<CallbackService>().set<CallbackService::Id::Stop>([this]() {
+      LOG(info) << "Saving records into ROOT file";
+      mAlign.terminate();
+    });
+  }
+
+  //_________________________________________________________________________________________________
+  void finaliseCCDB(framework::ConcreteDataMatcher& matcher, void* obj)
+  {
+    /// finalize the track extrapolation setting
+    if (mCCDBRequest && base::GRPGeomHelper::instance().finaliseCCDB(matcher, obj)) {
+      if (matcher == framework::ConcreteDataMatcher("GLO", "GRPMAGFIELD", 0)) {
+        TrackExtrap::setField();
+        TrackExtrap::useExtrapV2();
+        mAlign.SetBFieldOn(mch::TrackExtrap::isFieldON());
+        trackFitter.smoothTracks(true);
+      }
+
+      if (matcher == framework::ConcreteDataMatcher("GLO", "GEOMALIGN", 0)) {
+        LOG(info) << "Loading reference geometry from CCDB";
+        transformation = geo::transformationFromTGeoManager(*gGeoManager);
+        for (int i = 0; i < 156; i++) {
+          int iDEN = GetDetElemId(i);
+          transform[iDEN] = transformation(iDEN);
+        }
+      }
+    }
+  }
+
+  //_________________________________________________________________________________________________
+  void run(framework::ProcessingContext& pc)
+  {
+    if (mCCDBRequest) {
+      base::GRPGeomHelper::instance().checkUpdates(pc);
+    }
+
+    o2::globaltracking::RecoContainer recoData;
+    recoData.collectData(pc, *mDataRequest.get());
+
+    const auto& mchTracks = recoData.getMCHTracks();
+    const auto& mchmidMatches = recoData.getMCHMIDMatches();
+    const auto& mchClusters = recoData.getMCHTrackClusters();
+
+    for (auto const& mchmidMatch : mchmidMatches) {
+
+      int mchTrackID = mchmidMatch.getMCHRef().getIndex();
+      const auto& mchTrack = mchTracks[mchTrackID];
+      int first = mchTrack.getFirstClusterIdx();
+      int last = mchTrack.getLastClusterIdx();
+      int Ncluster = last - first + 1;
+
+      if (Ncluster <= 9) {
+        continue;
+      }
+
+      mch::Track convertedTrack;
+
+      for (int i = first; i <= last; i++) {
+        const auto& cluster = mchClusters[i];
+        convertedTrack.createParamAtCluster(cluster);
+      }
+
+      // Erase removable track
+      if (!RemoveTrack(convertedTrack)) {
+        mAlign.ProcessTrack(convertedTrack, transformation, false, weightRecord);
+        pc.outputs().snapshot(Output{"MUON", "RECORD_MCHMID", 0}, *mAlign.GetRecord());
```

Comment:
I asked @ktf , the DPL in this case will create a multip-part message with each snapshot being a single part. On the receiving side you would need to explicitly do something like
```
auto pos = inputs.getPos(binding);
size_t nparts = inputs.getNofParts(pos);
for (size_t i=0;I<nparts;i++) {
   auto mpRecord = inputs.get<fwdalign::MillePedeRecord>(binding, I);
   // ... use the record
}
```
With the default `inputs.get<>(binding)` you will get only 1st out of potentially many tracks. I think this is what RootTreeWriter is using. You should probably check if the number of tracks you write in this device is equal to those in the written file.

Reply:
Thanks a lot for info!
With my current version of sending tracks to output(using `LifeTime::TimeFrame`), I've got actually a total number of entries for my output record file, that is compatible with what I've got using AO2D. And the content inside this root file seems fine to me as well cuz I can actually run my alignment workflow with this record file as input, and I didn't get any trouble...

I will check if the number of tracks that are being processed corresponds to the number of entries of my output record file. I will let you know very shortly.

Reply:
HI,
I tested already with `Lifetime::Sporadic`: I actually have ~40 tracks candidates per `run`, and the number of total tracks candidates that were being converted into records corresponds to the number of total entries I have in my output ROOT file of `MillePede` records. I may stay with this solution which leads me to the smallest modifications compared to the solution proposed by Giulio.